be/src/vec/functions/function_soundex.cpp

```diff
@@ -0,0 +1,119 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cctype>
+
+#include "common/status.h"
+#include "vec/columns/column_string.h"
+#include "vec/data_types/data_type_string.h"
+#include "vec/functions/function.h"
+#include "vec/functions/simple_function_factory.h"
+
+namespace doris::vectorized {
+#include "common/compile_check_begin.h"
+
+class FunctionSoundex : public IFunction {
+public:
+    static constexpr auto name = "soundex";
+
+    static FunctionPtr create() { return std::make_shared<FunctionSoundex>(); }
+
+    String get_name() const override { return name; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeString>();
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        uint32_t result, size_t input_rows_count) const override {
+        const ColumnPtr col_ptr = block.get_by_position(arguments[0]).column;
+
+        auto res_column = ColumnString::create();
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            StringRef ref = col_ptr->get_data_at(i);
+            std::string soundex_code = calculate_soundex(ref);
+            res_column->insert_data(soundex_code.c_str(), soundex_code.length());
```

Comment:
you can reserve a max size to avoid frequent malloc. then in calculation you can directly insert data to result column.